### ISSUE TITLE
Update Wikipedia Citation Template to meet guidelines

### ISSUE THREAD

**Issue by aurimasv** [zotero/translators#797]
See http://en.wikipedia.org/wiki/Wikipedia:Citation_templates and https://forums.zotero.org/discussion/41819/wikipedia-citation-templates-and-their-limitations/#Item_6 (but there might be more changes that need to be made)

**Comment by owcz** [comment]
@aurimasv: I can help with this but since it's been a year and I don't know what has been updated, might be useful to define the spec

**Comment by adam3smith** [comment]
The main job is actually tracking down the current version of the WP citation template and details about preferences for the various fields, which you are probably in a better position to do than us. 

The actual edits should be fairly light and just keep with the current structure of the translator (note that you can't use Scaffold for export translators, so you'll just have to test this in Zotero proper.

**Comment by owcz** [comment]
Yep, I'm really familiar with the citation templates—my question is just what specifically needs updating (specific fields for specific item types, I'd presume?)

I know, for instance, that the citation template chokes on `YYYY-MM` formats (which JSTOR is wont to give) and prefers `MMMM YYYY`

The other thing here is that some users will prefer having the citation templates formed with ref tags around them (`<ref name="xxxx">{{cite book ... }}</ref>`) rather than the `{{cite book}}` alone. And some users will also prefer either mdy or dmy formatted dates over the `YYYY-MM-DD` notation. It would be nice to let users select these options but it looks like there isn't a mechanism for that... and we'd end up just offering slightly modified styles off of the default

**Comment by adam3smith** [comment]
Yeah, we won't offer customization (it technically is possible to code translator-specific preferences into translators that can be changed eiterh via checkbox or hidden preference, but we discourage it whenever there isn't a very strong case). So what we're after is, in this order of priority:
1) Something that works reliably with WP and
2) Wherever there is something like a standard for WP, follows that.
3) Where 1) and 2) don't apply, follow general standards. E.g. prefer ISO dates (YYYY-MM-DD) over other formats. 

The list at the top of 
https://forums.zotero.org/discussion/41819/wikipedia-citation-templates-and-their-limitations with following discussion would be a good starting point. The discussion also gives a good sense about the types of considerations. What we'd obviously want to avoid is too much of a favor of your (or anoyone else's) personal preferences, though I understand if you'll have to make judgment calls in some places (ideally the template would be welll defined enough so you wouldn't, but the way I understand it, that's not the case).

**Comment by owcz** [comment]
Where can I find the current Wikipedia CSL style? So I have a "Wikipedia Citation Templates" style in my Settings > Export > Quick Copy pane but there is no similar option in Settings > Cite > Styles or online at [zotero.com/styles](https://www.zotero.org/styles/?q=wikipedia). There are defaults now (e.g., [`{{cite book}}`](https://en.wikipedia.org/wiki/Template:Cite_book)) so there shouldn't be any major issues cleaning it up.

**Comment by adam3smith** [comment]
It's not a citation style (those aren't flexible enough), but an export translator. It's on github at 
https://github.com/zotero/translators/blob/master/Wikipedia%20Citation%20Templates.js
and as `Wikipedia Citation Template.js` in the translators folder of your [Zotero data directory](https://www.zotero.org/support/zotero_data). 
Export translators are only [minimally documented](https://www.zotero.org/support/dev/translators/coding#export_translators) so it's mainly straight javascript, learning by doing, and, in the beginning, sticking closely to the template.

**Comment by owcz** [comment]
@adam3smith Question about how date ranges are handled: a site's RIS might provide `Y1  - 2002/03//
`, which Zotero will convert to `2002-03`. Ostensibly this means "March 2002" but it can also mean "2002–2003". WP currently marks `2002-03` as an error because of this potential ambiguity. So my course of action is likely to convert these formats to MMMM YYYY on export, but shouldn't Zotero be distinguishing between the two ambiguities itself? You can see other date errors [here](https://en.wikipedia.org/wiki/Help:CS1_errors#Check_date_values_in:_.7Cparam1.3D.2C_.7Cparam2.3D.2C_...)

**Comment by adam3smith** [comment]
What matters for us is that 2002-03 isn't ambiguous for the purpose of Zotero's date parser (not least because Zotero doesn't understand date ranges at all at this time), which is why we use it. That also means that you can reliably export this correctly to whichever format Wikipedia prefers.

I'm actually not 100% firm on the details of how dates are exposed from the database to export translators. It [looks like](https://github.com/zotero/translators/blob/master/BibTeX.js#L1272) you actually have to run `ZU.strToDate(item.date)`, which will return an object with `.year` `.month` and `.day` but unless @aurimasv or @zuphilip happen to know this, you might have to experiment.

**Comment by aurimasv** [comment]
I'd look at what RIS does. It should handle dates fairly well

**Comment by zuphilip** [comment]
1) The [specification for RIS](http://refman.com/training/ris-format) for the field `DA` are clear on this:
<blockquote>Dates must be in the following format: 
YYYY/MM/DD/other info</blockquote>


2) I think that `item.date` will just return the string which then can be passed to the function [`ZU.strToDate`](https://github.com/zotero/zotero/blob/4.0/chrome/content/zotero/xpcom/date.js#L230) as @adam3smith pointed out. This function will return an object like:
```js
date = {
   order : 'ymd'
   year : 2016,
   month : 1,
   date : 23,
   parts : ...
}
```
For me, it looks that the function covers a lot of cases it is trying really hard to extract the parts of the date correctly. Thus, it makes normally sense to use this function for an export translator instead of the string value only.

**Comment by aurimasv** [comment]
I meant the RIS export translator above

**Comment by zuphilip** [comment]
@aurimasv I guess that would be https://github.com/zotero/translators/blob/master/RIS.js#L1902 and following lines?

**Comment by owcz** [comment]
I spent a few hours on this today and it's looking really good. I'm just cleaning it up and running some more test cases now.

I'd like to build in Wikidata support via [maxlath/wikidata-sdk](../../../maxlath/wikidata-sdk). For instance, on Wikipedia export at least, I'd like to search Wikidata/Wikipedia for whether an article exists on "The American Political Science Review" or "Giovanni Baldelli" so I can include a link in the citation. Where is the best place to put the dependent JS and what's the best way to integrate it into (or call it from) Wikipedia_Citation_Templates.js? Would this functionality be useful for any other translator/exporter?

**Comment by zuphilip** [comment]
The translator documentation is partially outdated and maybe a little chaotic, but it is a wiki, where everybody can help to improve this state, e.g. the function list https://www.zotero.org/support/dev/translators/functions

Your work on this translator sounds great and I am curious to see a PR. I understand that you included a Wikidata lookup, which uses the above mentioned JS files. I do something [similiar](https://github.com/UB-Mannheim/zotkat/blob/master/PicaSWB.js#L86-98) in one of my costumized non-official translator (this is currently mainly a proof of concept). However, I don't think we do this in any (official) export translator at the moment. Maybe we have to think whether in the normal zotero workflow this makes sense at all to lookup more information during export. But I think these things are easier to discuss if we see some code.

**Comment by owcz** [comment]
@zuphilip I believe the wiki is restricted—whenever I log in, I can't edit. I haven't added Wikidata lookup yet but I'd like to try. I was looking for some advice on how best to include/call external scripts (where they should live, etc.) Also if this kind of stuff is verboten I can just save myself the time. The primary use case is to check whether a work or author mentioned in a citation (e.g., "Dissent" magazine in `|work=Dissent`) has its own WP page such that we can automatically link it in the citation (e.g., `|work=[[Dissent (American magazine)|Dissent]]`). I think it'd be useful.

**Comment by zuphilip** [comment]
I would then suggest that you finish your work on the wikipedia export translator without any special lookup. In the meantime we can discuss further if any lookup and external js-library would be okay. I am skeptical but some other people should say a word as well.

**Comment by dstillman** [comment]
>  I believe the wiki is restricted—whenever I log in, I can't edit.

You have to request editing access by emailing support@zotero.org with your wiki username.

**Comment by adam3smith** [comment]
agree with @zuphilip -- let's leave lookup out for now. Discuss on Zotero forums or zotero-dev, but like him, I'm skeptical.

**Comment by owcz** [comment]
I'll leave it out of the PR, but just to be clear, when you both say "skeptical"—about what?

**Comment by adam3smith** [comment]
I'm skeptical about the overhead that entails, including performance implications, and I'm skeptical about including external queries in an export translator, which no other such translator does. 
But skeptical means just that; I'm open to be convinced.

**Comment by owcz** [comment]
Several months ago, there appears to have been a change in how the export translator pulled from the active Zotero selection. For example, if several entries were highlighted, they would export in that order, from top to bottom, but now they appear to be out of order or in some random order. Is there a different function that we should be using now to preserve the selection order?

**Comment by adam3smith** [comment]
@owcz -- I'm looking through old issues: did you ever turn this into a PR? Can the current template export be improved?

**Comment by owcz** [comment]
Nope and absolutely! Hard to believe it's been five years but the one I wrote then is still going strong. I'd need to pull some personal touches out of it and add coverage for the rarer citation templates but should be a good weekend project some time in the new year.

**Comment by imsys** [comment]
I just tried Zotero and I really like it! For my use case I'm just missing the extra identifiers like pmc, pmid, s2cid (semanticscholar), biorxiv, arxiv, etc. I was able to add all those fields by tweaking [wikipedia-templates.csl](https://github.com/citation-style-language/styles/blob/master/wikipedia-templates.csl), but I'm having problems making CSL understand the S2CID variable, I'm guessing it doesn't accept numbers in the name.
